### PR TITLE
fix `hash` and `==` for LibGit2.Oid

### PR DESCRIPTION
**Pull request by JeffBezanson**
I happened to find these minor infelicities in the libgit2 code.

### CODE REVIEW

File: base/libgit2/oid.jl [hunk 1]
```diff
@@ -71,12 +71,12 @@ Base.string(id::Oid) = hex(id)
 
 Base.show(io::IO, id::Oid) = print(io, "Oid($(string(id)))")
 
-Base.hash(id::Oid) = hash(id.val)
+Base.hash(id::Oid, h::UInt) = hash(id.val, h)
 
-cmp(id1::Oid, id2::Oid) = int(ccall((:git_oid_cmp, :libgit2), Cint,
-                                         (Ptr{Oid}, Ptr{Oid}), Ref(id1), Ref(id2)))
```
Comment (yuyichao) [inline base/libgit2/oid.jl]: So it seems that this is not tested?

Comment (JeffBezanson) [inline base/libgit2/oid.jl]: I guess so. Perhaps hashing and comparing these Oids is not even necessary; I'm hoping somebody who knows this code better can comment.

**Comment by StefanKarpinski** [comment]
Food for thought on protocols. With implicit protocol satisfaction, this would silently fail to implement the `Hashable` protocol. With explicit statement, you could get an error that the protocol isn't satisfied. Either one would be better than this where it silently doesn't work, although I'm inclined to feel that the explicit case is better.

**Comment by JeffBezanson** [comment]
Point taken. However some careful design is needed. For example, I might implement `hash` and `==` methods, but forget e.g. to write `<: Hashable` on the type declaration, in which case you also lose checking. Defining methods inside protocol blocks can prevent this, but then you can't introduce a protocol that describes methods somebody else already implements without going back and editing their code. So I don't really know the answer yet.

**Comment by JeffBezanson** [comment]
I actually think just having

```
protocol Hashable
    hash(_, UInt)
    ==(_, _)
end
```

written down somewhere helps a lot with this. The docs for `hash` also describe this, but that doesn't seem to work as well psychologically. I can think of many reasons why this might be the case.

**Comment by StefanKarpinski** [comment]
Yeah, I'm still pretty up in the air about which is better myself. I'm slightly inclined towards explicit opt-in.

**Comment by JeffBezanson** [comment]
I'm fine with an explicit version if we can come up with a nice design.

To me, the key is that being able to ask whether type X implements a protocol is strictly separate from *wanting* X to implement the protocol. So to me it would be a shame if the core design tied these together.

**Comment by StefanKarpinski** [comment]
True, those are definitely distinct. Do you think it's a good idea to allow the programmer to express that they want to implement a protocol without implementing it? If so, which one does dispatch pay attention to?

**Comment by JeffBezanson** [comment]
The best thing I can think of currently is (1) build in the fully-implicit version, (2) be able to write something like `@implements Array Iterable` that will check `Array <: Iterable` and issue diagnostics. I realize this is exactly what I criticized above as insufficiently automatic to prevent problems like the one here with LibGit2.Oid, but it's the best I have so far.

File: base/libgit2/oid.jl (hunk 1 above)
Comment (jakebolewski) [inline base/libgit2/oid.jl]: We currently use the bytestring SHA's everywhere, but that should be transitioned over to using Oid's for some type safety.

**Pull request merged**